app/src/main/java/com/onewelcome/showcaseapp/feature/userregistration/twostepregistration/TwoStepRegistrationViewModel.kt

```diff
@@ -0,0 +1,220 @@
+import com.onegini.mobile.sdk.android.model.entity.UserProfile
+import com.onewelcome.core.omisdk.handlers.CreatePinRequestHandler
+import com.onewelcome.core.omisdk.handlers.TwoStepRegistrationRequestHandler
+import com.onewelcome.core.usecase.GetTwoStepIdentityProvidersUseCase
+import com.onewelcome.core.usecase.GetUserProfilesUseCase
+import com.onewelcome.core.usecase.IsSdkInitializedUseCase
+import com.onewelcome.core.usecase.StatelessUserRegistrationUseCase
+import com.onewelcome.core.usecase.UserRegistrationUseCase
+import com.onewelcome.core.util.Constants
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class TwoStepRegistrationViewModel @Inject constructor(
+  isSdkInitializedUseCase: IsSdkInitializedUseCase,
+  private val userRegistrationUseCase: UserRegistrationUseCase,
+  private val statelessUserRegistrationUseCase: StatelessUserRegistrationUseCase,
+  private val getTwoStepIdentityProvidersUseCase: GetTwoStepIdentityProvidersUseCase,
+  private val getUserProfilesUseCase: GetUserProfilesUseCase,
+  private val twoStepRegistrationRequestHandler: TwoStepRegistrationRequestHandler,
+  private val createPinRequestHandler: CreatePinRequestHandler,
+) : ViewModel() {
+
+  var uiState by mutableStateOf(State())
+    private set
+
+  private val _navigationEvents = Channel<NavigationEvent>(Channel.BUFFERED)
+  val navigationEvents = _navigationEvents.receiveAsFlow()
+
+  init {
+    viewModelScope.launch {
+      isSdkInitializedUseCase.execute().let { uiState = uiState.copy(isSdkInitialized = it) }
+      updateIdentityProviders()
+      updateSelectedIdentityProvider()
+      updateUserProfiles()
+      updateCancellationButton()
+    }
+  }
+
+  private fun enterOptionalData() {
+    viewModelScope.launch {
+      _navigationEvents.send(NavigationEvent.ToTwoStepInputScreen)
+    }
+  }
+
+  fun onEvent(event: UiEvent) {
+    when (event) {
+      is UiEvent.StartTwoStepRegistration -> startRegistration()
+      is UiEvent.UpdateSelectedIdentityProvider -> uiState = uiState.copy(selectedIdentityProvider = event.identityProvider)
+      is UiEvent.UpdateSelectedScopes -> uiState = uiState.copy(selectedScopes = event.scopes)
+      is UiEvent.CancelRegistration -> cancelRegistration()
+      is UiEvent.SetStatelessRegistration -> uiState = uiState.copy(isStatelessRegistration = event.isStateless)
+      is UiEvent.SubmitResponseCode -> submitResponseCode(event.responseCode)
+      is UiEvent.UpdateResponseCode -> uiState = uiState.copy(responseCode = event.code)
+      is UiEvent.SubmitOptionalData -> subnitOptionalData(event.optionalData)
+      is UiEvent.EnterTwoStepInput -> enterOptionalData()
+    }
+  }
+
+  private fun subnitOptionalData(optionalData: String) {
+    twoStepRegistrationRequestHandler.optionalData = optionalData
```

Comment:
let's have a public method in twoStepRegistrationRequestHandler to update optionalData. Make optionalData variable in TwoStepRegistrationRequestHandler private.